server/schedule/operator.go

```diff
@@ -104,6 +106,50 @@ func (rp RemovePeer) Influence(opInfluence OpInfluence, region *core.RegionInfo)
 	from.RegionCount--
 }
 
+// MergeRegion is an OperatorStep that merge two regions.
+type MergeRegion struct {
+	FromRegion uint64
+	ToRegion   uint64
+	Direction  metapb.MergeDirection
+	// there are two regions involved in merge process,
+	// so to keep them from other scheduler,
+	// both of them should add MerRegion operatorStep.
+	// But actually, tikv just need the region want to be merged to get the merge request,
+	// thus use a IsPssive mark to indicate that
+	// this region doesn't need to send merge request to tikv.
+	IsPassive   bool
+	OldStartKey []byte
+	OldEndKey   []byte
+}
+
+func (mr MergeRegion) String() string {
+	return fmt.Sprintf("merge region %v into region %v", mr.FromRegion, mr.ToRegion)
+}
+
+// IsFinish checks if cuurent step is finished
+func (mr MergeRegion) IsFinish(region *core.RegionInfo) bool {
+	if mr.Direction == metapb.MergeDirection_Forward {
+		return bytes.Compare(region.Region.StartKey, mr.OldStartKey) < 0
+	}
+	// for the case that the region endkey change to "" after merge
+	// note that "" is smaller than any key, comapring to old endkey it will not regared as finished.
+	// so it need a special judge
+	if bytes.Compare(region.Region.EndKey, []byte("")) == 0 && bytes.Compare(mr.OldEndKey, []byte("")) != 0 {
+		return true
+	}
+	return bytes.Compare(region.Region.EndKey, mr.OldEndKey) > 0
```

Comment:
I think `mr.OldEndKey` can never be nil, so how about
`return len(region.Region.EndKey) == 0 || bytes.Compare(region.Region.EndKey, mr.OldEndKey) > 0`